src/steth/scDAI.sol

```diff
@@ -0,0 +1,533 @@
+        uint256 _totalCollateral,
+        uint256 _invested,
+        uint256 _sDaiNeeded
+    ) internal {
+        // handle rounding errors when withdrawing everything
+        _sDaiNeeded = _sDaiNeeded > _totalCollateral ? _totalCollateral : _sDaiNeeded;
+        // to keep the same ltv, total debt in weth to be repaid has to be proportional to total sDai collateral we are withdrawing
+        uint256 wethNeeded = _sDaiNeeded.mulDivUp(_totalDebt, _totalCollateral);
+        wethNeeded = wethNeeded > _invested ? _invested : wethNeeded;
+
+        uint256 wethDisinvested = 0;
+        if (wethNeeded != 0) wethDisinvested = _disinvest(wethNeeded);
+
+        // repay debt and withdraw collateral from each protocol in proportion to sDai supplied
+        uint256 length = protocolAdapters.length();
+
+        for (uint256 i = 0; i < length; i++) {
+            (uint256 id, address adapter) = protocolAdapters.at(i);
+            uint256 collateral = IAdapter(adapter).getCollateral(address(this));
+
+            if (collateral == 0) continue;
+
+            uint256 debt = IAdapter(adapter).getDebt(address(this));
+            uint256 toWithdraw = _sDaiNeeded.mulDivUp(collateral, _totalCollateral);
+
+            if (wethDisinvested != 0 && debt != 0) {
+                // keep the same ltv when withdrawing sDai supplied from each protocol
+                uint256 toRepay = toWithdraw.mulDivUp(debt, collateral);
+
+                if (toRepay > wethDisinvested) {
+                    toRepay = wethDisinvested;
+                } else {
+                    wethDisinvested -= toRepay;
+                }
+
+                _repay(id, toRepay);
+            }
+
+            _withdraw(id, toWithdraw);
+        }
+    }
+
+    function _calculateTotalAssets(uint256 _float, uint256 _collateral, uint256 _invested, uint256 _debt)
+        internal
+        view
+        returns (uint256 total)
+    {
+        total = _float + _collateral;
+
+        uint256 profit = _calculateWethProfit(_invested, _debt);
+
+        if (profit != 0) {
+            // account for slippage when selling weth profits
+            total += priceConverter.ethTosDai(profit).mulWadDown(slippageTolerance);
+        } else {
+            total -= priceConverter.ethTosDai(_debt - _invested);
+        }
+    }
+
+    function _calculateWethProfit(uint256 _invested, uint256 _debt) internal pure returns (uint256) {
+        return _invested > _debt ? _invested - _debt : 0;
+    }
+
+    function _wethBalance() internal view returns (uint256) {
+        return weth.balanceOf(address(this));
+    }
+
+    function _swapWethForAsset(uint256 _wethAmount, uint256 _sDaiAmountOutMin) internal returns (uint256) {
+        // weth => usdc => dai
+        bytes memory result = address(swapper).functionDelegateCall(
+            abi.encodeWithSelector(Swapper.swapWethToSdai.selector, _wethAmount, _sDaiAmountOutMin)
+        );
+
+        return abi.decode(result, (uint256));
+    }
```

Comment:
while trying to optimize the rebalance script for scUSDC for a specific case where unrealized profits are less than newly deposited assets i realized it cannot be done essentially because this function is internal. So you should make it external and onlyKeeper for scDAI so we can avoid the same situation here. 

To elaborate further, selling profits for scUSDC means to redeem some shares from scETH and use the returned amount of WETH for WETH -> USDC swap so the yield is converted to USDC. If there are also new deposits included in this case, we want to avoid redeeming scETH shares if not necessary ie the unrealized profit can be "covered" with new deposits thus no redeem call is needed. But as I realized recently, this is not possible because the mechanism of swapping WETH to USDC is internally restricted.

Reply:
if we make it onlyKeeper then the withdraw function fails, because this is being used in the withdraw method too when users withdraw funds

Reply:
but that's not really an issue, you can just wrap this function inside a new "swapWethForAsset" public function which will include the _onlyKeeper check

Reply:
Oh yes. Nice. Thank you so much. [Done](https://github.com/lindy-labs/sandclock-contracts/pull/155/commits/9bea4d805b2b8a4fac2925173976b4fbec3fe05d)